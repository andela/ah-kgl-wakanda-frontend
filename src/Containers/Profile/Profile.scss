@import '../../assets/sass/global.scss';
#profile {
  input,
  button {
    width: 80%;
  }
  button {
    width: 80%;
  }
  label {
    color: #444;
  }
  textarea {
    border: 2px solid #efefef;
    margin-bottom: 10px;
    padding: 0 6px;
    width: 79%;
    background: #fafafa;
    &:focus {
      outline: none !important;
      border: 2px solid rgba(#aca9a9, 0.5);
      box-shadow: 2px 0 6px rgba(#efefef, 0.8);
    }
  }
  .navbar {
    background-color: white;
  }
  .sidenav {
    height: 90%;
  }
  .bodyProfile {
    height: 200%;
  }
}

.profile {
  padding-left: 14%;

  div,
  .profile {
    background-color: #f5f5f5;
  }
  height: 100%;
  #pic {
    img {
      width: 70%;
    }
    .uname {
      padding: 50px 10;
    }
  }
}
@media only screen and (max-width: 700px) {
  #profile {
    padding-left: 10%;
    #pic {
      display: flex;
      height: 100px;
      .form-box {
        width: 40%;
        height: 0;
        padding-top: 0%;
        img {
          width: 110%;
        }
      }

      .uname {
        padding: 0;
      }
    }
  }
  input,
  .button {
    width: 90% !important;
  }
  textarea {
    width: 90% !important;
    margin: 5px 0;
    padding: 3px;
  }
}
.p-avatar.big {
  margin-left: 80%;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 120px;
  height: 120px;
  border-radius: 50%;

  .avatar-main {
    width: 100%;
    height: 100%;
    border-radius: 50%;
    box-shadow: 0 2px 5px rgba(0, 0, 0, 0.12);
    background-size: cover;
    margin-bottom: 20%;
  }
  .edit {
    background-color: #f46036;
    border-radius: 50%;
    position: absolute;
    height: 28px;
    width: 28px;
    margin-left: 6.8rem;
    margin-top: 1.6rem;
    text-align: center;
    padding: 4px;
    color: white;
    font-size: 14px;
    cursor: pointer;
  }
}
.col-sm-3.form-box.uname {
  margin-top: 6%;
  margin-left: 14%;
}

.col-sm-10.body.bodyProfile {
  padding-bottom: 0px;
  height: 100%;
}

.body.profile {
  height: 10%;
  div {
    margin-top: 2%;
    margin-bottom: 2%;
    padding: 0;
  }
}
@media only screen and (max-width: 300px) {
  #profile {
    padding-top: 0px;
    padding-left: 10%;
    #pic {
      display: flex;
      height: 120px;
      .form-box {
        padding-top: 0%;
        img {
          width: 70%;
        }
      }

      .uname {
        padding: 0;
      }
    }
  }
  input,
  .button {
    width: 90% !important;
  }
  textarea {
    width: 90% !important;
    margin: 5px 0;
    padding: 3px;
  }
}
@media only screen and (max-width: 500px) {
  #profile {
    padding-top: 0px;
    padding-left: 10%;
    #pic {
      display: flex;
      height: 120px;
      .p-avatar.row.big {
        margin-left: -23%;
        margin-top: 40%;
      }
      .form-box {
        padding-top: 0%;
        img {
          width: 70%;
        }
      }

      .uname {
        padding: 0;
        margin-top: 27%;
        margin-left: 15%;
      }
    }
  }
  input,
  .button {
    width: 90% !important;
  }
  textarea {
    width: 90% !important;
    margin: 5px 0;
    padding: 3px;
  }
  .form.profile {
    margin-top: 23%;
    margin-right: 12%;
  }
}

.form-box {
  padding-top: 0%;
  background: #f5f5f5;
}
