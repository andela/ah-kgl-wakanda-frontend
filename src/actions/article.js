/* eslint no-use-before-define: 0 */
import { toast } from 'react-toastify';
import {
  PUBLISH_ARTICLE,
  GET_SINGLE_ARTICLE,
  DELETE_SINGLE_ARTICLE,
  LOAD,
  STOP_LOADING,
  DISPLAY_COMMENTS,
} from '../actionTypes/article';
import { NOT_FOUND, SUCCESS_MESSAGE } from '../actionTypes/system';
import wakanda from '../api/wakanda';

/**
 *Update comment
 *
 * @param {*} {slug}
 * @param {*} id
 * @returns {void}
 */
export const updateComment = ({ slug, id, body }) => async () => {
  try {
    await wakanda.put(`/api/articles/${slug}/comments/${id}`, {
      comment: {
        body,
      },
    });

    return true;
  } catch (error) {
    if (error.response) {
      toast.error(error.response.data.message);
    } else {
      toast.error('Internet Lost');
    }

    return false;
  }
};

/**
 *Delete comment
 *
 * @param {*} {slug}
 * @param {*} id
 * @returns {void}
 */
export const deleteComment = ({ slug, id }) => async () => {
  try {
    await wakanda.delete(`/api/articles/${slug}/comments/${id}`);

    return true;
  } catch (error) {
    if (error.response) {
      toast.error(error.response.data.message);
    } else {
      toast.error('Internet Lost');
    }

    return false;
  }
};

/**
 * Fetch comments
 * @return {object} response
 * @param {String} slug
 */
export const fetchComments = slug => dispatch => {
  wakanda
    .get(`/api/articles/${slug}/comments`)
    .then(response => {
      const {
        data: { data: comments },
      } = response;

      dispatch({
        type: DISPLAY_COMMENTS,
        payload: comments,
      });
    })
    .catch(error => {
      if (error.response) {
        toast.error(error.response.data.message);
      } else {
        toast.error('Internet Lost');
      }
    });
};

/**
 *Posts a comment on an article
 *
 * @param {*} { slug, comment }
 * @returns {void}
 */
export const commentArticle = ({ slug, body }) => async dispatch => {
  dispatch({
    type: LOAD,
  });

  if (body.length === 0) {
    dispatch({
      type: STOP_LOADING,
    });
    return;
  }

  try {
    await wakanda.post(`/api/articles/${slug}/comments`, {
      comment: {
        body,
      },
    });

    toast.success('Comment posted');
    dispatch({
      type: STOP_LOADING,
    });

    return true;
  } catch (error) {
    dispatch({
      type: STOP_LOADING,
    });

    if (error.response) {
      toast.error(error.response.data.message);
    } else {
      toast.error('Internet Lost');
    }

    return false;
  }
};

/**
 * @returns {*} dispatch
 * @param {string} article
 */
export default article => dispatch => {
  wakanda
    .post('/api/articles', {
      article,
    })
    .then(response => {
      const newArticle = response.data.data;
      dispatch({
        type: PUBLISH_ARTICLE,
        payload: newArticle.article,
      });
    })
    .catch(error => {
      if (error.response) {
        toast.error(error.response.data.message);
      } else {
        toast.error('Internet Lost');
      }
    });
};

/**
 * @returns {*} dispatch
 * @param {string} slug
 * @param {string} article
 */
export const updateArticle = (slug, article) => dispatch => {
  wakanda
    .put(`/api/articles/${slug}`, {
      article,
    })
    .then(response => {
      const updatedArticle = response.data.data;
      dispatch({
        type: PUBLISH_ARTICLE,
        payload: updatedArticle.article,
      });
    })
    .catch(error => {
      if (error.response) {
        toast.error(error.response.data.message);
      } else {
        toast.error('Internet Lost');
      }
    });
};

/**
 * @returns {*} dispatch
 * @param {string} slug
 */
export const getArticle = slug => dispatch => {
  wakanda
    .get(`/api/articles/${slug}`)
    .then(async response => {
      const { article } = response.data.data;

      await fetchComments(slug);

      dispatch({
        type: GET_SINGLE_ARTICLE,
        payload: article,
      });
    })
    .catch(error => {
      if (error.response && error.response.status === 404) {
        const payload = {
          notFound: {
            status: true,
            attribute: 'Article',
          },
        };
        dispatch({
          type: NOT_FOUND,
          payload,
        });
      }
    });
};

/**
 * @returns {*} dispatch
 * @param {string} slug
 */
export const deleteArticle = slug => dispatch => {
  wakanda
    .delete(`/api/articles/${slug}`)
    .then(response => {
      const { message } = response.data;
      const successNotification = {
        successMessage: {
          status: true,
          message,
        },
      };
      dispatch({
        type: DELETE_SINGLE_ARTICLE,
      });
      dispatch({
        type: SUCCESS_MESSAGE,
        payload: successNotification,
      });
    })
    .catch(error => {
      if (error.response) {
        toast.error(error.response.data.message);
      } else {
        toast.error('Internet Lost');
      }
    });
};

/**
 * Likes an article
 *
 * @param {*} slug
 * @returns {void}
 */
export const likeArticle = slug => async dispatch => {
  try {
    await wakanda.post(`/api/articles/${slug}/favorite`);
    toast.success('Liked');

    dispatch(getArticle(slug));

    return true;
  } catch (error) {
    if (error.response) {
      if (error.response.data.status === 409) {
        dispatch(unlikeArticle(slug));
        return;
      }
      toast.error('You must be logged in');
    } else {
      toast.error('Internet Lost');
    }
  }
};

/**
 * Unlike an article
 *
 * @param {*} slug
 * @returns {void}
 */
export const unlikeArticle = slug => async dispatch => {
  try {
    await wakanda.delete(`/api/articles/${slug}/favorite`);
    toast.success('Unliked');

    dispatch(getArticle(slug));

    return false;
  } catch (error) {
    if (error.response) {
      if (error.response.data.status === 409) {
        dispatch(likeArticle(slug));
        return;
      }
      toast.error('You must be logged in');
    } else {
      toast.error('Internet Lost');
    }
  }
};
