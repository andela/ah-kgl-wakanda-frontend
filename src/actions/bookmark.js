import { toast } from 'react-toastify';
import wakanda from '../api/wakanda';
import { BOOKMARK_SUCCESS, BOOKMARK_STARTED, BOOKMARK_ERROR } from '../actionTypes/bookmark';

/**
 * bookmark action start
 * @returns {object} action
 */
export const bookmarkStarted = () => ({
  type: BOOKMARK_STARTED,
});

/**
 * on bookmark action success
 * @param {object} data
 * @returns {object} action
 */
export const bookmarkSuccess = data => ({
  type: BOOKMARK_SUCCESS,
  payload: data,
});

/**
 * on bookmark action error
 * @param {*} error
 * @returns {object} action
 */
export const bookmarkError = error => ({
  type: BOOKMARK_ERROR,
  payload: error,
});

/**
 * action bookmark on articles
 * @param {string} slug
 * @returns {func} dispatch
 */
export const bookmark = slug => async dispatch => {
  dispatch(bookmarkStarted());
  try {
    const res = await wakanda.post(`/api/articles/${slug}/bookmark`);
    dispatch(bookmarkSuccess(res.data.message));
    toast.success(res.data.message);

    return res;
  } catch (e) {
    const error = e.response.data
      ? e.response.data.message
      : 'Please check your internet connection';
    dispatch(bookmarkError(error));
    toast.error(error);
    return e;
  }
};
