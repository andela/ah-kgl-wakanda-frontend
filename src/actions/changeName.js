import { CHANGE_NAME } from '../actionTypes';

/**
 * @returns {*} dispatch
 * @param {string} name
 */
export const changeName = name => dispatch => {
  // Do some actions
  dispatch({
    type: CHANGE_NAME,
    payload: name,
  });
};
