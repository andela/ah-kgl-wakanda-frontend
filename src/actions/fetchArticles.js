import axios from 'axios';
import {
  FETCH_ARTICLES_STARTED,
  FETCH_ARTICLES_SUCCESS,
  FETCH_ARTICLES_ERROR,
  FETCH_USER_ARTICLES,
  NEW_FEED,
  NEW_USER_ARTICLES,
} from '../actionTypes/fetchArticles';
import wakanda from '../api/wakanda';

/**
 * Triggers an action
 * when the fetching process starts
 * @return {object} action
 */
export const fetchStart = () => ({
  type: FETCH_ARTICLES_STARTED,
});

/**
 * Triggers an action
 * when the fetching process succeeds
 * @param {object} articles
 * @return {object} action
 */
export const fetchSuccess = articles => ({
  type: FETCH_ARTICLES_SUCCESS,
  payload: articles,
});

/**
 * Triggers an action
 * when the fetching process succeeds
 * @param {object} articles
 * @return {object} action
 */
export const fetchNewFeedSuccess = articles => ({
  type: NEW_FEED,
  payload: articles,
});

/**
 * Triggers an action
 * when the fetching process succeeds
 * @param {object} articles
 * @return {object} action
 */
export const fetchUserArticlesSuccess = articles => ({
  type: FETCH_USER_ARTICLES,
  payload: articles,
});

/**
 * Triggers an action
 * when the fetching process succeeds
 * @param {object} articles
 * @return {object} action
 */
export const fetchNewUserArticlesSuccess = articles => ({
  type: NEW_USER_ARTICLES,
  payload: articles,
});

/**
 * Triggers an action
 * when the fetching process fails
 * @param {object} error
 * @return {object} action
 */
export const fetchError = error => ({
  type: FETCH_ARTICLES_ERROR,
  payload: error,
});

/**
 * Fetch comments
 * @return {object} response
 * @param {String} slug
 */
export const fetchComments = async slug => {
  const res = await axios.get(
    `${process.env.REACT_APP_BACKEND_URL}/api/articles/${slug}/comments`,
    {
      headers: {
        'Content-Type': 'application/json',
      },
    },
  );

  const {
    data: {
      data: { commentsCount },
    },
  } = res;

  return commentsCount;
};

/**
 * Gets user articles
 * @param {string} username
 * @returns {void}
 */
export const fetchUserArticles = username => async dispatch => {
  dispatch(fetchStart());

  try {
    const res = await axios.get(
      `${process.env.REACT_APP_BACKEND_URL}/api/articles/${username}/private?limit=10`,
      {
        headers: {
          'Content-Type': 'application/json',
        },
      },
    );

    const {
      data: {
        data: { articles },
      },
    } = res;

    if (articles.length === 0) {
      dispatch(fetchUserArticlesSuccess(articles));
      return;
    }

    // Map articles with comments
    await articles.forEach(async (element, index) => {
      const count = await fetchComments(element.slug);
      articles[index].commentsCount = count;
      if (index === articles.length - 1) dispatch(fetchUserArticlesSuccess(articles));
    });
  } catch (e) {
    if (e.response) {
      dispatch(fetchError(e.response.data.message));
    }
    dispatch(fetchError('Please check your internet connection'));
  }
};
/**
 * Gets user articles
 * @param {string} username
 * @param {string} offset
 * @param {string} limit
 * @returns {void}
 */
export const fetchNewUserArticles = (username, offset, limit) => async dispatch => {
  dispatch(fetchStart());

  try {
    const res = await axios.get(
      `${process.env.REACT_APP_BACKEND_URL}/api/articles/${username}/private?offset=${offset}&limit=${limit}`,
      {
        headers: {
          'Content-Type': 'application/json',
        },
      },
    );

    const {
      data: {
        data: { articles },
      },
    } = res;

    if (articles.length === 0) dispatch(fetchNewUserArticlesSuccess(articles));

    // Map articles with comments
    await articles.forEach(async (element, index) => {
      const count = await fetchComments(element.slug);
      articles[index].commentsCount = count;
      if (index === articles.length - 1) dispatch(fetchNewUserArticlesSuccess(articles));
    });
  } catch (e) {
    if (e.response) {
      dispatch(fetchError(e.response.data.message));
    }
    dispatch(fetchError('Please check your internet connection'));
  }
};

/**
 * Fetch articles
 * @return {object} response
 */
export const fetchArticles = () => async dispatch => {
  dispatch(fetchStart());

  try {
    const res = await axios.get(`${process.env.REACT_APP_BACKEND_URL}/api/articles?limit=10`, {
      headers: {
        'Content-Type': 'application/json',
      },
    });

    const {
      data: { articles },
    } = res.data;

    // Map articles with comments
    await articles.forEach(async (element, index) => {
      const count = await fetchComments(element.slug);
      articles[index].commentsCount = count;
      if (index === articles.length - 1) dispatch(fetchSuccess(articles));
    });
  } catch (e) {
    if (e.response) {
      dispatch(fetchError(e.response.data.message));
    }
    dispatch(fetchError('Please check your internet connection'));
  }
};

/**
 * @returns {*} dispatch
 * @param {number} offset
 * @param {number} limit
 */
export const fetchNewFeed = (offset, limit) => dispatch => {
  dispatch(fetchStart());
  wakanda
    .get(`/api/articles?offset=${offset}&limit=${limit}`)
    .then(async response => {
      const { articles } = response.data.data;

      if (articles.length === 0) dispatch(fetchNewFeedSuccess(articles));
      // Map articles with comments
      await articles.forEach(async (element, index) => {
        const count = await fetchComments(element.slug);
        articles[index].commentsCount = count;
        if (index === articles.length - 1) dispatch(fetchNewFeedSuccess(articles));
      });
    })
    .catch(e => {
      if (e.response) {
        dispatch(fetchError(e.response.data.message));
      }
      dispatch(fetchError('Please check your internet connection'));
    });
};
