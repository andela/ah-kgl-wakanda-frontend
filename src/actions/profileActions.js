import * as types from '../actionTypes/profileActionTypes';
import wakanda from '../api/wakanda';
/**
 * @returns {*} object
 * @param {string} payload
 */
export const viewProfile = payload => dispatch => {
  return wakanda
    .get(`api/users/${payload}`)
    .then(res => {
      const { profile } = res.data;
      dispatch({
        type: types.SET_PROFILE,
        payload: profile,
      });
    })
    .catch(err => {
      err.message = 'Please check your internet and reload';
      document.getElementById('message').innerHTML = err.message;
    });
};

/**
 * @returns {*} object
 * @param {string} payload
 * @param {string} username
 */
export const editProfile = (payload, username) => dispatch => {
  wakanda.put(`api/user/${username}`, { ...payload }).then(res => {
    const { profile } = res.data;
    viewProfile();
    dispatch({
      type: types.EDIT_PROFILE,
      payload: profile,
    });
  });
};
