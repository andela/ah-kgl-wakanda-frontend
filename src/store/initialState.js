const initialState = {
  name: { team: 'Wakanda' },
  title: 'Hello world',
};

export default initialState;
