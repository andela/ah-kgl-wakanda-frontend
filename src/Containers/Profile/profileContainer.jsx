import axios from 'axios';
import React, { Component } from 'react';
import { PropTypes } from 'prop-types';
import { connect } from 'react-redux';
import { ToastContainer } from 'react-toastify';
import Input from '../../Components/Common/input/input';
import { viewProfile, editProfile } from '../../actions/profileActions';
import './Profile.scss';
import NavBar from '../../Components/NavBar/NavBar';
import SideBar from '../../Components/SideBar/SideBar';
import picture from '../../assets/img/blank_profile_pic.png';

const CLOUDINARY_URL = 'https://api.cloudinary.com/v1_1/ah-wakanda/image/upload';
const CLOUDINARY_UPLOAD_PRESET = 'j9eazq6w';

/**
 *
 * @param {event} e
 * @returns {method} render
 */
export class ProfileContainer extends Component {
  state = {
    bio: '',
  };

  fileInput = React.createRef();

  /**
   *
   * @returns {method} form
   */
  componentWillMount() {
    const { onViewProfile, username } = this.props;
    onViewProfile(username);
  }

  onKeyChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  edit = e => {
    e.preventDefault();
    const { onEditProfile, onViewProfile, username } = this.props;
    onEditProfile(this.state, username);
    onViewProfile(username);
  };

  onUpload = e => {
    const file = e.target.files[0];
    this.setState({ [e.selectedFile]: e.target.files[0] });
    const formData = new FormData();
    formData.append('file', file);
    formData.append('upload_preset', CLOUDINARY_UPLOAD_PRESET);
    axios({
      url: CLOUDINARY_URL,
      method: 'POST',
      header: {
        'Content-Type': 'application/x-ww-form-urlencoded',
      },
      data: formData,
    })
      .then(res => {
        const { user } = this.props;
        user.image = res.data.url;
        this.setState({ user });
        const { onEditProfile } = this.props;
        onEditProfile(user, user.username);
      })
      .catch(err => {
        err.message = 'Please check your internet and reload';
        document.getElementById('message').innerHTML = err.message;
      });
  };

  /**
   * @param {object} props
   * @returns {method} form
   */
  form = () => {
    const { user } = this.props;
    const { bio } = this.state;
    return (
      <form action="#" className="form profile" onSubmit={this.edit}>
        <div className="row">
          <div className="col-sm-5 form-box">
            <Input
              type="text"
              name="firstname"
              label="Frist Name"
              defaultValue={user.firstname || ''}
              placeholder=""
              onType={this.onKeyChange}
            />
          </div>
          <div className="col-sm-5 form-box">
            <Input
              type="text"
              name="lastname"
              label="last Name"
              placeholder=""
              defaultValue={user.lastname || ''}
              onType={this.onKeyChange}
            />
          </div>
        </div>
        <div className="row">
          <div className="col-sm-5 form-box">
            <label htmlFor="bio">
              Bio
              <br />
              <textarea
                id="bio"
                rows="4"
                name="bio"
                cols="39"
                value={bio || user.bio}
                onChange={this.onKeyChange}
              />
            </label>
          </div>
          <div className="col-sm-5 form-box">
            <Input
              type="email"
              name="email"
              label="Email"
              defaultValue={user.email || ''}
              placeholder=""
              onType={this.onKeyChange}
            />
            <button className="button button-register">Edit</button>
          </div>
        </div>
      </form>
    );
  };

  /**
   *
   * @returns {jsx} react fragment
   */
  render() {
    const { user } = this.props;
    return (
      <React.Fragment>
        <ToastContainer
          position="top-right"
          autoClose={3000}
          hideProgressBar={false}
          newestOnTop
          closeOnClick
          rtl={false}
          pauseOnVisibilityChange
          draggable
          pauseOnHover
        />
        <div id="profile">
          <div className="wrapper">
            <div className="row">
              <NavBar
                {...this.props}
                username={user.username}
                picture={user.username}
                currentUser={user}
                displaySearchBox={false}
              />

              <div className="col-sm-2 sideB">
                <SideBar user={user} />
              </div>
              <div className="col-sm-10 form-box bodyProfile">
                <div className="body profile">
                  <div id="pic" className="row">
                    <div className="col-sm-2 form-box pl-4">
                      <div className="p-avatar row big">
                        <div
                          className="avatar-main"
                          style={{
                            backgroundImage: user.image ? `url(${user.image})` : `url(${picture})`,
                          }}
                        >
                          <div
                            className="edit"
                            role="presentation"
                            onClick={() => this.fileInput.current.click()}
                          >
                            <i className="fas fa-pencil-alt" />
                          </div>
                        </div>

                        <input
                          type="file"
                          name="file"
                          className="file-upload"
                          onChange={this.onUpload}
                          ref={this.fileInput}
                          style={{ display: 'none' }}
                        />
                      </div>
                    </div>
                    <div className="col-sm-3 form-box uname">
                      <b>{`${user.firstname || ''} ${user.lastname || ''}`}</b>
                      <br />
                      {`@${user.username}`}
                    </div>
                  </div>
                  {this.form()}
                  <p id="message" />
                </div>
              </div>
            </div>
          </div>
        </div>
      </React.Fragment>
    );
  }
}

ProfileContainer.defaultProps = {
  onViewProfile: null,
  onEditProfile: null,
  username: '',
};

ProfileContainer.propTypes = {
  onViewProfile: PropTypes.object,
  user: PropTypes.object.isRequired,
  username: PropTypes.string,
  onEditProfile: PropTypes.object,
};
/**
 *
 * @param {object} state
 * @param {object} currentUser
 * @return {void}
 */
export const mapStateToProps = ({
  profile: { user },
  currentUser: {
    user: { username },
  },
}) => ({ user, username });

/**
 *
 * @param {object} dispatch
 * @returns {method} dispatch
 */
export const mapDispatchToProps = dispatch => ({
  onViewProfile: username => dispatch(viewProfile(username)),
  onEditProfile: (profile, username) => dispatch(editProfile(profile, username)),
});
export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ProfileContainer);
