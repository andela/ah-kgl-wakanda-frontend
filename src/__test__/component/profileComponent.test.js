import React from 'react';
import { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import mockAxios from 'axios';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { ProfileContainer, mapDispatchToProps } from '../../Containers/Profile/profileContainer';
import initialState from '../../store/initialState';
import { viewProfile, editProfile } from '../../actions/profileActions';

const e = { preventDefault: jest.fn() };
const props = {
  currentUser: initialState.currentUser,
  profile: {
    user: {
      username: 'giselei',
      firstname: 'abana',
      lastname: 'ni beza',
      email: 'gisele.iradukunda@andela.com',
      bio: 'cyane man',
      image:
        'http://res.cloudinary.com/ah-wakanda/image/upload/v1561649492/ax27ojnfvsgmoosgcgyi.jpg',
      follows: 0,
      followings: 0,
      articles: 0,
      allowEmailNotification: false,
    },
  },
  onViewProfile: jest.fn(),
  onEditProfile: jest.fn(e),
  location: {
    pathname: '/api/user/giselei',
  },
};
const { profile } = props;
const mockStore = configureMockStore([thunk]);
const store = mockStore(initialState);
let wrapper;
describe('Renders view Profile Component', () => {
  it('Should render the <ProfileContainer/>', () => {
    wrapper = shallow(
      <Provider store={store}>
        <ProfileContainer {...profile} />
      </Provider>,
    );
    expect(wrapper).toMatchSnapshot();
    wrapper.find('input[name="email"]');
    wrapper.find('input[name="firstname"]');
    wrapper.find('input[name="lastname"]');
    wrapper.find('form');
  });
  it('Should map dispatch to props', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).onViewProfile(profile.user.username);
    expect(dispatch).toHaveBeenCalled();
  });
  it('Should map dispatch to props', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).onEditProfile(profile.user.username);
    expect(dispatch).toHaveBeenCalled();
  });
});
describe('View & edit profile actions', () => {
  it('should successfully call the viewProfile action', async () => {
    mockAxios.get.mockImplementationOnce(() =>
      Promise.resolve({
        data: {
          status: 200,
          profile: {
            username: 'giselei',
            firstname: 'abana',
            lastname: 'ni beza',
            email: 'gisele.iradukunda@andela.com',
            bio: 'cyane man',
            image:
              'http://res.cloudinary.com/ah-wakanda/image/upload/v1561649492/ax27ojnfvsgmoosgcgyi.jpg',
            follows: 0,
            followings: 0,
            articles: 0,
          },
        },
      }),
    );
    await store.dispatch(viewProfile(profile.user.username));
  });
  it('should successfully call the editProfile action', async () => {
    mockAxios.put.mockImplementationOnce(() =>
      Promise.resolve({
        data: {
          status: 200,
          profile: {
            username: 'giselei',
            firstname: 'abana',
            lastname: 'ni beza',
            email: 'gisele.iradukunda@andela.com',
            bio: 'cyane man',
            image:
              'http://res.cloudinary.com/ah-wakanda/image/upload/v1561649492/ax27ojnfvsgmoosgcgyi.jpg',
            follows: 0,
            followings: 0,
            articles: 0,
          },
        },
      }),
    );
    await store.dispatch(editProfile(profile, profile.user.username));
  });
});
