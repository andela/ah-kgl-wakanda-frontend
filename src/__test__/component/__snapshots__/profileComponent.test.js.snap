// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders view Profile Component Should render the <ProfileContainer/> 1`] = `ShallowWrapper {}`;
